packages/flutter/lib/src/material/dialog.dart

```diff
@@ -972,7 +972,7 @@ Future<T?> showDialog<T>({
   )
   Widget? child,
 }) {
-  assert(child == null || builder == null);
+  assert(child != null || builder != null);
```

Comment:
I don't think this change achieves what you're looking for. The original assert simply doesn't allow for both to be null at the same time, which is valid use case: We don't want users to define both, even though the code technically allows for it. 

I think what you meant to do was introduce another assert:
```suggestion
  assert(child == null || builder == null);
  assert(child != null || builder != null);
```

We can need add a test to ensure that the assert is triggered to make sure there are no regressions

Reply:
I helped write the tests for you since it can be quite convoluted to write assertion tests:
```dart
  testWidgets('showDialog child and builder cannot be simultaneously defined', (WidgetTester tester) async {
    late BuildContext currentBuildContext;
    await tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          body: Center(
            child: Builder(
              builder: (BuildContext context) {
                currentBuildContext = context;
                return Container();
              }
            ),
          ),
        ),
      ),
    );

    expect(() async {
      showDialog<void>(
        context: currentBuildContext,
        child: const Text('Child'),
        builder: (BuildContext context) {
          return const Text('Builder');
        },
      );
    }, throwsAssertionError);
  });

  testWidgets('showDialog child or builder must be defined', (WidgetTester tester) async {
    late BuildContext currentBuildContext;
    await tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          body: Center(
            child: Builder(
              builder: (BuildContext context) {
                currentBuildContext = context;
                return Container();
              }
            ),
          ),
        ),
      ),
    );

    expect(
      () => showDialog<void>(context: currentBuildContext),
      throwsAssertionError,
    );
  });
```